### PR TITLE
refactor hier_part_rule abs path

### LINKED ISSUE #636: `uri_rule` requires absolute path
The test

```
BOOST_TEST(parse_uri("http://u:p@a.b:a").has_value());
```

works but it shouldn't.

It should require a `'/'` after the authority (after `:`)

```
   hier-part     = "//" authority path-abempty
                 / path-absolute
                 / path-rootless
                 / path-empty
   path-abempty  = *( "/" segment )
```

but it works because it considers ":" to be a valid `port-part` and `a` to be an independent valid path.

### PR DESCRIPTION
**Pull request by alandefreitas**
fix #636

### CODE REVIEW

File: include/boost/url/rfc/detail/impl/hier_part_rule.ipp [hunk 1]
```diff
@@ -65,7 +65,13 @@ parse(
...
         t.has_authority = true;
     }
-    if(it == end)
+    // the authority requires an absolute path
+    // or an empty path
+    if(it == end ||
+       (t.has_authority &&
+         (*it != '/' &&
+          *it != '?' &&
+          *it != '#')))
```
Comment (vinniefalco) [inline include/boost/url/rfc/detail/impl/hier_part_rule.ipp:74]: needs a test or two obviously

Comment (vinniefalco) [reply include/boost/url/rfc/detail/impl/hier_part_rule.ipp:74]: The spacing is weird, I would have written it as
```
    if(it == end || (
        t.has_authority && (
            *it != '/' &&
            *it != '?' &&
            *it != '#')))
```

This way everything is aligned on tab stops without added spaces

Comment (alandefreitas) [reply include/boost/url/rfc/detail/impl/hier_part_rule.ipp:74]: Oh! :open_mouth: Done.

I forgot to bring any tests from the `format` branch. I came up with a simpler test here.

Comment (vinniefalco) [reply include/boost/url/rfc/detail/impl/hier_part_rule.ipp]: I like simpler tests

**Comment by alandefreitas** [comment]
Is this good to go then?

**Comment by vinniefalco** [comment]
Yep.

**Pull request merged**